### PR TITLE
fix: ensure JWT segments have the right types

### PR DESCRIPTION
**Pull request by ret2libc**
Function `_unverified_decode` assumes `header` is a dict and performs a `.get("alg")` after decoding the JWT token, however the header may be a different object which does not have the `get` method. Similar problem for the `payload` segment, which is accessed as a dictionary but there is no check for it.

This PR fixes the problem by checking the type of `header`/`payload` in the `_unverified_decode` function.

### CODE REVIEW

File: tests/test_jwt.py [hunk 1]
```diff
@@ -126,6 +126,23 @@ def test_decode_valid(token_factory):
     assert payload["metadata"]["meta"] == "data"
 
 
+def test_decode_header_object(token_factory):
+    payload = token_factory()
+    payload = b"M7." + b".".join(payload.split(b".")[1:])
```
Comment (clundin25) [inline tests/test_jwt.py:131]: Can you add a comment that explains this line?

File: tests/test_jwt.py [hunk 2]
```diff
@@ -126,6 +126,23 @@ def test_decode_valid(token_factory):
...
+
+    with pytest.raises(ValueError) as excinfo:
+        jwt.decode(payload, certs=PUBLIC_CERT_BYTES)
+    assert excinfo.match(r"Header segment should be a JSON object: b'M7'")
+
+
+def test_decode_payload_object(signer):
+    payload = jwt.encode(signer, "iatexp")
```
Comment (clundin25) [inline tests/test_jwt.py:139]: Please add a comment to this as well, and how it correlates to the error message

**Comment by clundin25** [comment]
Minor comments, otherwise LGTM.

Thank you for your contribution!

**Review by clundin25 — APPROVED** [review]

File: google/auth/jwt.py [hunk 3]
```diff
@@ -133,11 +133,12 @@ def _unverified_decode(token):
         token (Union[str, bytes]): The encoded JWT.
 
     Returns:
-        Tuple[str, str, str, str]: header, payload, signed_section, and
+        Tuple[Mapping, str, str, str]: header, payload, signed_section, and
```
Comment (sai-sunder-s) [inline google/auth/jwt.py:136]: the payload is Mapping as well?

Comment (ret2libc) [reply google/auth/jwt.py]: Yes, fixed!

**Comment by clundin25** [comment]
Hi @ret2libc,

In order for the tests to pass can you merge https://github.com/clundin25/google-auth-library-python/commit/17cc33f139c18cef29db8f38e335dc1f3c476b45? Alternatively, I can push it to your branch if you give me write permissions to your fork.

Thanks!

**Comment by ret2libc** [comment]
> Hi @ret2libc,
> 
> In order for the tests to pass can you merge [clundin25@17cc33f](https://github.com/clundin25/google-auth-library-python/commit/17cc33f139c18cef29db8f38e335dc1f3c476b45)? Alternatively, I can push it to your branch if you give me write permissions to your fork.
> 
> Thanks!

Done!

File: tests/test_jwt.py [hunk 4]
```diff
@@ -288,9 +309,9 @@ def test_decode_no_key_id(token_factory):
 
 
 def test_decode_unknown_alg():
-    headers = json.dumps({u"kid": u"1", u"alg": u"fakealg"})
+    headers = json.dumps({"kid": "1", "alg": "fakealg"})
```
Comment (clundin25) [inline tests/test_jwt.py:312]: Are these changes necessary? We still unofficially support Python2 to a degree.

Comment (ret2libc) [reply tests/test_jwt.py:312]: No they aren't but I think they were generated by the formatter suggested in the project.

Comment (clundin25) [reply tests/test_jwt.py:312]: Interesting, thank you

Comment (clundin25) [reply tests/test_jwt.py:312]: Looks like this causes the Python2.7 unit tests to fail. Do you mind manually undoing this?

Comment (ret2libc) [reply tests/test_jwt.py]: Done!

**Comment by clundin25** [comment]
Hi @ret2libc,

Sorry for the delay, I was OOTO. Can you sync your branch? Sorry for the toil.

**Comment by ret2libc** [comment]
@clundin25 done!

**Comment by clundin25** [comment]
@ret2libc looks like there is a test failure.

```
    def test_decode_header_object(token_factory):
        payload = token_factory()
        # Create a malformed JWT token with a number as a header instead of a
        # dictionary (3 == base64d(M7==))
        payload = b"M7." + b".".join(payload.split(b".")[1:])

        with pytest.raises(ValueError) as excinfo:
            jwt.decode(payload, certs=PUBLIC_CERT_BYTES)
>       assert excinfo.match(r"Header segment should be a JSON object: b'M7'")
E       AssertionError: Pattern "Header segment should be a JSON object: b'M7'" not found in 'Header segment should be a JSON object: M7'
```

Also, just in case, here is a fresh refresh token for the system tests: https://github.com/clundin25/google-auth-library-python/commit/d246cfee23bc6e9694e325a96e578c6b37cb079e

**Comment by clundin25** [comment]
Hi @ret2libc,

This PR is getting bogged down by the test token used by the system tests expiring before getting this merged.

I have posted an updated token on my account and linked to it above. 

Alternatively you could give me write permissions to your fork and I will get this PR wrapped up.

Thank you again for this contribution!

**Comment by ret2libc** [comment]
> Hi @ret2libc,
> 
> This PR is getting bogged down by the test token used by the system tests expiring before getting this merged.
> 
> I have posted an updated token on my account and linked to it above.
> 
> Alternatively you could give me write permissions to your fork and I will get this PR wrapped up.
> 
> Thank you again for this contribution!

Sorry for being slow on this! Hopefully it works now. If it doesn't, let me know which commit to include and I'll try to do it ASAP.

**Comment by clundin25** [comment]
No need to apologize @ret2libc, we appreciate your contribution and I apologize for the system test churn.

Here is a fresh token. https://github.com/googleapis/google-auth-library-python/commit/b691524ca1d6db82816ffdc3ebdee15c4f834c24

**Comment by ret2libc** [comment]
Oh, I see. It's python2. Will have a fix by EOD.

**Comment by ret2libc** [comment]
@clundin25 do you have any idea what's going on now?

**Comment by clundin25** [comment]
Sorry for the delay. @ret2libc looks like the formatter choked.

This patch should resolve it.

https://github.com/clundin25/google-auth-library-python/commit/bc02d36576cf195c686571a2e71bb2858dab5fc0

**Comment by clundin25** [comment]
Thanks again for your contribution!
